File: 2019/intcode/intcode.ts
// Old code runs fine with ts-node but repeatedly ran into errors
// when compiling with tsc. Turns out it just really hates extending
// builtin classes like Map. `Memory` implemented another way
// to get around this.

class Memory {
  map: Map<number, number>;
  constructor(entries: [number, number][]) {
    this.map = new Map(entries);
  }

  set(k: number, v: number) {
    this.map.set(k, v);
  }

  get(k: number): number {
    return this.map.get(k) ?? 0;
  }
}

export type IntcodeGenerator = Generator<number, number, number>;

function* intcodeGenerator(prog: number[]): IntcodeGenerator {
  function getParams(
    program: Memory,
    parameterModes: number[],
    nParams: number,
    i: number,
    relativeBase: number
  ): number[] {
    const parameters: number[] = [];
    for (let j = 0; j < nParams; j++) {
      if (parameterModes[j]) {
        if (parameterModes[j] === 1) {
          // immediate mode
          parameters.push(i + j + 1);
        } else if (parameterModes[j] === 2) {
          // relative mode
          parameters.push(program.get(i + j + 1) + relativeBase);
        }
      } else {
        // position mode
        parameters.push(program.get(i + j + 1));
      }
    }
    return parameters;
  }

  const program = new Memory(prog.map((a, i) => [i, a]));
  let i = 0; // program index
  let relativeBase = 0;
  while (i >= 0) {
    const opcode = program.get(i) % 100;
    const parameterModes = Math.trunc(program.get(i) / 100)
      .toString()
      .split("")
      .reverse()
      .map(Number);

    let params: number[] = [];
    switch (opcode) {
      case 5:
      case 6:
        params = getParams(program, parameterModes, 2, i, relativeBase).map(
          (a) => program.get(a)
        );
        break;
      case 1:
      case 2:
      case 7:
      case 8:
        params = getParams(program, parameterModes, 3, i, relativeBase);
        break;
      case 3:
      case 4:
      case 9:
        params = getParams(program, parameterModes, 1, i, relativeBase);
        break;
    }

    switch (opcode) {
      case 1:
        // add
        program.set(params[2], program.get(params[0]) + program.get(params[1]));
        i += 3;
        break;
      case 2:
        // multiply
        program.set(params[2], program.get(params[0]) * program.get(params[1]));
        i += 3;
        break;
      case 3:
        // save input
        program.set(params[0], yield -9999);
        i++;
        break;
      case 4:
        //output
        yield program.get(params[0]);
        i++;
        break;
      case 5:
        // jump if true
        if (params[0] !== 0) {
          i = params[1];
          continue;
        } else {
          i += 2;
        }
        break;
      case 6:
        // jump if false
        if (params[0] === 0) {
          i = params[1];
          continue;
        } else {
          i += 2;
        }
        break;
      case 7:
        // less than
        if (program.get(params[0]) < program.get(params[1])) {
          program.set(params[2], 1);
        } else {
          program.set(params[2], 0);
        }
        i += 3;
        break;
      case 8:
        // equal
        if (program.get(params[0]) === program.get(params[1])) {
          program.set(params[2], 1);
        } else {
          program.set(params[2], 0);
        }
        i += 3;
        break;
      case 9:
        // adjust relative base
        relativeBase += program.get(params[0]);
        i++;
        break;
      case 99:
        // halt
        i = -99;
        break;
      default:
        throw new Error("Unknown opcode: " + program.get(i));
    }
    i++;
  }
  return program.get(0);
}

export default intcodeGenerator;

File: 2019/intcode/Intcode.go
package intcode

import (
	"fmt"
	"math"
)

type IntcodeProgram struct {
	Program      map[int]int
	programIndex int
	relativeBase int
	ch           chan int
}

const (
	REQUESTSIGNAL = math.MaxInt
	ENDSIGNAL     = math.MinInt
	STOPSIGNAL    = math.MinInt
)

func NewIntcodeProgram(prog []int, ch chan int) *IntcodeProgram {
	program := make(map[int]int)
	for i, n := range prog {
		program[i] = n
	}
	ic := &IntcodeProgram{program, 0, 0, ch}
	return ic
}

func (ic *IntcodeProgram) getParams(parameterModes []int, nParams int) []int {
	parameters := make([]int, nParams)
	for j := range nParams {
		if j >= len(parameterModes) || parameterModes[j] == 0 {
			// position mode
			parameters[j] = ic.Program[ic.programIndex+j+1]
		} else if parameterModes[j] == 1 {
			// immediate mode
			parameters[j] = ic.programIndex + j + 1
		} else if parameterModes[j] == 2 {
			// relative move
			parameters[j] = ic.Program[ic.programIndex+j+1] + ic.relativeBase
		}
	}
	return parameters

}

func (ic *IntcodeProgram) Run() {
	defer close(ic.ch)
	for ic.programIndex >= 0 {
		opcode := ic.Program[ic.programIndex] % 100
		parameterModeNumber := ic.Program[ic.programIndex] / 100
		parameterModes := make([]int, 0, 4)
		for parameterModeNumber > 0 {
			parameterModes = append(parameterModes, parameterModeNumber%10)
			parameterModeNumber /= 10
		}

		var params []int
		switch opcode {
		case 5:
			fallthrough
		case 6:
			tmp := ic.getParams(parameterModes, 2)
			params = make([]int, len(tmp))
			for i, n := range tmp {
				params[i] = ic.Program[n]
			}
		case 1:
			fallthrough
		case 2:
			fallthrough
		case 7:
			fallthrough
		case 8:
			params = ic.getParams(parameterModes, 3)
		case 3:
			fallthrough
		case 4:
			fallthrough
		case 9:
			params = ic.getParams(parameterModes, 1)
		}

		switch opcode {
		case 1:
			// add
			ic.Program[params[2]] = ic.Program[params[0]] + ic.Program[params[1]]
			ic.programIndex += 3
		case 2:
			// multiply
			ic.Program[params[2]] = ic.Program[params[0]] * ic.Program[params[1]]
			ic.programIndex += 3
		case 3:
			// save input
			ic.ch <- REQUESTSIGNAL
			received := <-ic.ch
			if received == STOPSIGNAL {
				return
			}
			ic.Program[params[0]] = received
			ic.programIndex++
		case 4:
			// output
			ic.ch <- ic.Program[params[0]]
			ic.programIndex++
		case 5:
			// jump if true
			if params[0] != 0 {
				ic.programIndex = params[1]
				continue
			} else {
				ic.programIndex += 2
			}
		case 6:
			// jump if false
			if params[0] == 0 {
				ic.programIndex = params[1]
				continue
			} else {
				ic.programIndex += 2
			}
		case 7:
			// less than
			if ic.Program[params[0]] < ic.Program[params[1]] {
				ic.Program[params[2]] = 1
			} else {
				ic.Program[params[2]] = 0
			}
			ic.programIndex += 3
		case 8:
			// equal
			if ic.Program[params[0]] == ic.Program[params[1]] {
				ic.Program[params[2]] = 1
			} else {
				ic.Program[params[2]] = 0
			}
			ic.programIndex += 3
		case 9:
			// adjust relative base
			ic.relativeBase += ic.Program[params[0]]
			ic.programIndex++
		case 99:
			// halt
			ic.ch <- ENDSIGNAL
			return
		default:
			panic(fmt.Sprintf("Unknown opcode: %d", ic.Program[ic.programIndex]))
		}
		ic.programIndex++
	}

}
